File: front/types/course.ts
// frontend/types/course.ts
export interface Course {
    id: number;
    title: string;
    code: string;
    department: string;
    level: string;
    credits: number;
    hours: number;
    teacher_id: number | null;
    teacher_name: string;
    filiere_id: number | null;
    university_id: number;
  }
  
  export interface CourseHistory {
    id: number;
    course_id: number;
    action: 'created' | 'updated' | 'deleted';
    field_changed: string | null;
    old_value: string | null;
    new_value: string | null;
    user_name: string;
    created_at: string;
  }
  
  export interface CourseFormData {
    title: string;
    code: string;
    level: string;
    filiere_id: string;
    teacher_id: string;
    credits: string;
    hours: string;
  }
  
  export interface CourseFilters {
    search: string;
    level: string;
    filiere: string;
  }

File: back/app/models/course.py
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.database.connection import Base

class Course(Base):
    __tablename__ = "courses"
    
    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=False)
    code = Column(String, unique=True, nullable=False)
    department = Column(String, nullable=False)
    level = Column(String, nullable=False)
    teacher_id = Column(Integer, ForeignKey("teachers.id"), nullable=True)
    credits = Column(Integer, default=3)
    hours = Column(Integer, default=20)
    university_id = Column(Integer, ForeignKey("universities.id"), nullable=False)
    filiere_id = Column(Integer, ForeignKey("filieres.id"), nullable=True)  

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    university = relationship("University", back_populates="courses")
    teacher = relationship("Teacher", back_populates="courses")
    grades = relationship("Grade", back_populates="course", cascade="all, delete-orphan")
    filiere = relationship("Filiere")  
    history = relationship("CourseHistory", back_populates="course", cascade="all, delete-orphan")